File: cpp/Mod_sizeLogspace.cpp
#pragma once

#include "Mod_firstChild.cpp"
#include "Mod_hasNextSibling.cpp"
#include "Mod_nextSibling.cpp"

int sizeLogspace_debug(Tree t, Node n){
	cout << "action;node;count" << endl;
	Node start = n;
	int last = 0; //0-down, 1-over, 2-up
	int cnt = 1;
	if(isLeaf(t, n)){
		return cnt;
	}
	cout << "down;" << n << ";" << cnt << endl;
	n = firstChild(t, n);
	cnt++;
	
	while(true){
		if(last == 0 || last == 1){
			if(!isLeaf(t, n)){
				cout << "down;" << n << ";" << cnt << endl;
				n = firstChild(t, n);
				cnt++;
				last = 0;
			}else if(hasNextSibling(t, n)){
				cout << "over;" << n << ";" << cnt << endl;
				n = nextSibling(t, n);
				cnt++;
				last = 1;
			}else{
				cout << "up;" << n << ";" << cnt << endl;
				n = parent(t, n);
				if(n==start)
					return cnt;
				last = 2;
			}
		}else if(last == 2){
			if(hasNextSibling(t, n)){
				cout << "over;" << n << ";" << cnt << endl;
				n = nextSibling(t, n);
				cnt++;
				last = 1;
			}else{
				cout << "up;" << n << ";" << cnt << endl;
				n = parent(t, n);
				if(n == start)
					return cnt;
				last = 2;
			}
		}
	}	
}
int sizeLogspace(Tree t, Node n){
	Node start = n;
	int last = 0; //0-down, 1-over, 2-up
	int cnt = 1;
	if(isLeaf(t, n)){
		return cnt;
	}
	n = firstChild(t, n);
	cnt++;
	
	while(true){
		if(last == 0 || last == 1){
			if(!isLeaf(t, n)){
				n = firstChild(t, n);
				cnt++;
				last = 0;
			}else if(hasNextSibling(t, n)){
				n = nextSibling(t, n);
				cnt++;
				last = 1;
			}else{
				n = parent(t, n);
				if(n==start)
					return cnt;
				last = 2;
			}
		}else if(last == 2){
			if(hasNextSibling(t, n)){
				n = nextSibling(t, n);
				cnt++;
				last = 1;
			}else{
				n = parent(t, n);
				if(n == start)
					return cnt;
				last = 2;
			}
		}
	}	
}

File: hs/Mod_sizeLogspace.hs
module Mod_sizeLogspace ( sizeLogspace ) where
import Prelude hiding (and, or)
import Debug.Trace
import Splib
import Types
import Mod_firstChild
import Mod_hasNextSibling
import Mod_nextSibling

trace' :: String -> a -> a
--trace' str x = (trace str x) 
trace' str x = x

sizeLogspace tree node 
 | isLeaf tree node = 1
 | otherwise = trace' ("action;node;count\ndown;" ++ (show node) ++ ";" ++ (show 1)) (sizeHelper tree (firstChild tree node) 0 2 node)

sizeHelper t n lastA cnt start
 | n == start = cnt
 | downOrOver && (sp_not (isLeaf t n)) = trace' ("down;" ++ (show n) ++ ";" ++ show(cnt)) (sizeHelper t (firstChild t n) 0 (cnt+1) start)
 | downOrOver && hasNextSibling t n = trace' ("over;" ++ (show n) ++ ";" ++ show(cnt)) (sizeHelper t (nextSibling t n) 1 (cnt+1) start)
 | downOrOver = trace' ("up;" ++ (show n) ++ ";" ++ show(cnt)) (sizeHelper t (parent t n) 2 cnt start)
 | hasNextSibling t n = trace' ("over;" ++ (show n) ++ ";" ++ show(cnt)) (sizeHelper t (nextSibling t n) 1 (cnt+1) start)
 | otherwise = trace' ("up;" ++ (show n) ++ ";" ++ show(cnt)) (sizeHelper t (parent t n) 2 cnt start)
    where
     downOrOver = lastA == 0 || lastA == 1